spec/langchain/llm/openai_spec.rb

```diff
@@ -327,6 +327,35 @@
           subject.functions = [{foo: :bar}]
           expect(subject.chat(prompt: prompt, model: model, temperature: temperature)).to be_a Hash
         end
+
+        context "with function_calls responses in the history" do
```

Comment:
I'm not familiarized with the tool, but is the function response also added to the history after executing it on `langchainrb`? if so, maybe is worth to add a scenario that is supported that format?

I mean, openai api with function calling adds 2 new history message types, one when the model decides to call a function (the one you supported):
`{ role: "assistant", "content": null, "function_call" => "function_name" }`

But also you can add to the history an entry with `role: "function"` that has the response of the function call itself so the model knows the content of the function execution:

`{ role: "function", "content": "function execution response / response sent to the user/chat", "name": "function_name" }`

Reply:
AFAIK langchain.rb doesn't call functions automatically and adding that feature would mean a fair amount of work

Reply:
@andreibondarev if you want we can work in https://python.langchain.com/docs/modules/chains/how_to/openai_functions in another PR